File: scripts/Fraction to Recurring Decimal/Fraction to Recurring Decimal.java

class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if(numerator == 0){
            return  "0";
        }
        
        StringBuilder sb = new StringBuilder("");
        if(numerator<0 && denominator>0 || numerator>0 && denominator<0){
            sb.append("-");
        }
        
        long divisor = Math.abs((long)numerator);
        long dividend = Math.abs((long)denominator);
        long remainder = divisor % dividend;
        sb.append(divisor / dividend);
        
        if(remainder == 0){
            return sb.toString();
        }
        sb.append(".");
        HashMap<Long, Integer> map = new HashMap<Long, Integer>();
        while(remainder!=0){
            if(map.containsKey(remainder)){
                sb.insert(map.get(remainder), "(");
                sb.append(")");
                break;
            }
            map.put(remainder, sb.length());
            remainder*= 10;
            sb.append(remainder/dividend);
            remainder%= dividend;
        }
        return sb.toString();
    }
}

File: algorithms/F/Fraction to Recurring Decimal/Fraction to Recurring Decimal.js
var fractionToDecimal = function(numerator, denominator) {
    if(numerator == 0) return '0'
    let result = ''
    if(numerator*denominator <0){
        result += '-'
    }
    
    let dividend = Math.abs(numerator)
    let divisor = Math.abs(denominator)
    result += Math.floor(dividend/divisor).toString()
    
    let remainder = dividend % divisor
    if(remainder == 0) return result
    
    result += '.'
    
    let map1 = new Map()
    while(remainder != 0){
        if(map1.has(remainder)){
            let i = map1.get(remainder)
            result = result.slice(0, i) + '(' + result.slice(i) + ')'
            break;
        }
        map1.set(remainder, result.length)
        remainder *= 10
        result += Math.floor(remainder/divisor).toString()
        remainder %= divisor
    }
    return result
};
